resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java

```diff
@@ -1163,6 +1173,93 @@ public interface IFile extends IResource, IEncodedStorage, IAdaptable {
+	 * {@link #setContents(InputStream, boolean, boolean, IProgressMonitor)} with
+	 * {@code new ByteArrayInputStream(content)} as {@code InputStream}.
+	 *
+	 * @param content     the content bytes
+	 * @param force       a flag controlling how to deal with resources that are not
+	 *                    in sync with the local file system
+	 * @param keepHistory a flag indicating whether or not store the current
+	 *                    contents in the local history
+	 * @param monitor     a progress monitor, or <code>null</code> if progress
+	 *                    reporting is not desired
+	 * @throws CoreException if this method fails or is canceled.
+	 * @since 3.21
+	 */
+	public default void setContents(byte[] content, boolean force, boolean keepHistory, IProgressMonitor monitor)
+			throws CoreException {
+		int updateFlags = (force ? IResource.FORCE : IResource.NONE)
+				| (keepHistory ? IResource.KEEP_HISTORY : IResource.NONE);
+		setContents(content, updateFlags, monitor);
+	}
+
+	/**
+	 * Preferable equivalent of calling
+	 * {@link #setContents(InputStream, int, IProgressMonitor)} with
+	 * {@code new ByteArrayInputStream(content)} as {@code InputStream}.
+	 *
+	 * @param content     the content bytes
+	 * @param updateFlags bit-wise or of update flag constants (<code>FORCE</code>
+	 *                    and <code>KEEP_HISTORY</code>)
+	 * @param monitor     a progress monitor, or <code>null</code> if progress
+	 *                    reporting is not desired
+	 * @throws CoreException if this method fails or is canceled.
+	 * @since 3.21
+	 */
+	public default void setContents(byte[] content, int updateFlags, IProgressMonitor monitor) throws CoreException {
+		// Meant to be overridden for local files with Files.write
+		setContents(new ByteArrayInputStream(content), updateFlags, monitor);
+	}
+
+	/**
+	 * Reads the content in a byte array. This method is not intended for reading in
+	 * large files that do not fit in a byte array. Preferable equivalent of calling
+	 * {@code getContents(true).readAllBytes()}.
+	 *
+	 * @return content bytes
+	 * @throws CoreException on error
+	 * @see #getContents(boolean)
+	 * @since 3.21
+	 */
+	public default byte[] readAllBytes() throws CoreException {
+		// Meant to be overridden for local files
+		// as Files.readAllBytes is ~ 1.5 times faster
+		try (InputStream stream = getContents(true)) {
+			return stream.readAllBytes();
+		} catch (IOException e) {
+			throw new CoreException(Status.error("Error reading " + getFullPath(), e)); //$NON-NLS-1$
+		}
+	}
+
+	/**
+	 * Reads the content as char array. Skips the utf BOM header if any. This method
+	 * is not intended for reading in large files that do not fit in a char array.
+	 * Preferable equivalent of calling {@code readString().toCharArray()}
+	 *
+	 * @return content String without UTF BOM header
+	 * @throws CoreException on error
+	 * @see #readString()
+	 * @since 3.21
+	 */
+	public default char[] readAllChars() throws CoreException {
```

Comment:
Is this method that returns the in-between of a byte-array and a String really necessary?
AFAICT operating on char-arrays is not that common and there are more high-level and saver APIs like StringBuilder.
E.g. nio-Files also just offers `readAllBytes()` and `readString()/readLines()`.

In order to not blow up the API unnecessarily I think potentiall callers should be checked if they can work well with other the other methods.

Reply:
JDT requires char array. I don't like that but JDT has voted against changing everything to String. Converting from String to char[] is a memory and performance overhead. Converting from byte[] to char[] requires the Charset to be known, but that is currently only known by platform.

Reply:
Understand. Too bad.
Would it be possible to achieve the same performance by using only internals?
I'm asking because I'm not sure we should promote using char-arrays in general by adding corresponding APIs?

Reply:
I won't mind that additional method, it does not hurt to have i in the API as it is default implemented.

Reply:
> promote using char-arrays

There is nothing bad about returning arrays (same as byte array).

Reply:
> > promote using char-arrays
> 
> There is nothing bad about returning arrays (same as byte array).

Yes, the 'idea' was that for low level access there is the bytes method and otherwise the String.
But I understand the need for JDT and if everyone is fine with it, I'm fine as well.